### PR TITLE
task/FP-1947: Parameterized nginx `server_name`

### PR DESCRIPTION
**Pull request by rstijerina**
### Overview

This PR achieves two main goals:

1. Allows for sending in the `NGINX_SERVER_NAME` environment variable to redirect server web traffic (443/80) to a desired URL, e.g. a vanity URL like 3dem-staging.tacc.utexas.edu instead of the VM TACC hostname like 3dem04.tacc.utexas.edu
    a. This feature is built into the [official nginx docker image](https://hub.docker.com/_/nginx) (see Complex Configuration)
    b. There is now a `nginx/templates` dir where any file here ending in the `.template` suffix will be processed by nginx templating and will be deposited into `/etc/nginx/conf.d/`
2. Allows for portals to gain basic nginx functionality and settings with the `docker-compose.nginx.yml` file, without much of the cruft used specifically by cms/portal setups
    a. There is now a separate `nginx/nginx.conf` file which serves as a base for nginx with some preselected variables for use by any webserver
    b. There is now a separate `nginx/templates/default.conf.template` file which has portal-specific configuration

Much inspiration on settings adjustments gained from https://www.nginx.com/resources/wiki/start/topics/tutorials/config_pitfalls/

### Testing
1. Go to both http://3dem04.tacc.utexas.edu and https://3dem04.tacc.utexas.edu and confirm they both redirect to https://3dem-staging.tacc.utexas.edu

### Adoption

1. Before this PR is merged, you can set `CAMINO_VERSION` to `task/FP-1947--paramaterized-nginx-server-name` in the `*.env` file in the Core-Portal Deployments repo
2. Set `NGINX_SERVER_NAME` to the desired URL of the portal in the `*.env` file of the Core-Portal-Deployments repo

### Notes
**IMPORTANT**: On merge, `NGINX_SERVER_NAME` will be a **required** variable in the `*.env` file

### CODE REVIEW

**Review by taoteg — COMMENTED** [review]
Bueno! I like how you broke the default.conf template out! Clean =)

File: conf/nginx/nginx.conf [hunk 1]
```diff
@@ -1,128 +1,40 @@
+user nginx;
+worker_processes auto;
+
+
 events {
-  worker_connections  4096;
+    worker_connections 1024;
```
Comment (rstijerina) [inline conf/nginx/nginx.conf:6]: `worker_connections` is limited to the available simultaneous open files on the VM, which is actually `1024`, so no point in increasing beyond that:
```
[root@designsafe01 ~]# ulimit -a
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 63374
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1024                               <----- This determines upper limit of worker_connections
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 63374
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```

**Comment by taoteg** [comment]
Idea: If every deployment environment has to specify the `NGINX_SERVER_NAME`, and if NSO-Ops is going to be consistent about using the naming convention and location of `/etc/letsencrypt/live/hostanem/fullchain.pem` and `/etc/letsencrypt/live/hostname/privkey.pem`, then this portion of the `docker-compose.ENV.override.yml`:

```
    volumes:
      - /etc/letsencrypt/live/hostname/fullchain.pem:/etc/ssl/certs/portal.cer
      - /etc/letsencrypt/live/hostname/privkey.pem:/etc/ssl/private/portal.key
```

Could be moved into base `docker-compose.yml` as:

```
    volumes:
      - /etc/letsencrypt/live/${NGINX_SERVER_NAME}/fullchain.pem:/etc/ssl/certs/portal.cer
      - /etc/letsencrypt/live/${NGINX_SERVER_NAME}/privkey.pem:/etc/ssl/private/portal.key
```

and eliminate one more moving part.

**Comment by rstijerina** [comment]
> Idea: If every deployment environment has to specify the `NGINX_SERVER_NAME`, and if NSO-Ops is going to be consistent about using the naming convention and location of `/etc/letsencrypt/live/hostanem/fullchain.pem` and `/etc/letsencrypt/live/hostname/privkey.pem`, then this portion of the `docker-compose.ENV.override.yml`:
> 
> ```
>     volumes:
>       - /etc/letsencrypt/live/hostname/fullchain.pem:/etc/ssl/certs/portal.cer
>       - /etc/letsencrypt/live/hostname/privkey.pem:/etc/ssl/private/portal.key
> ```
> 
> Could be moved into base `docker-compose.yml` as:
> 
> ```
>     volumes:
>       - /etc/letsencrypt/live/${NGINX_SERVER_NAME}/fullchain.pem:/etc/ssl/certs/portal.cer
>       - /etc/letsencrypt/live/${NGINX_SERVER_NAME}/privkey.pem:/etc/ssl/private/portal.key
> ```
> 
> and eliminate one more moving part.

Eventually I think we could bring that in. Time being, however, it is pretty variable across environments

**Comment by wesleyboar** [comment]
I like the idea and the result. It's just a bit over my head to offer a professional critique though.

**Review by taoteg — APPROVED** [review]
Worked in ECEP Prod deployment. Great work!

**Pull request merged**